### PR TITLE
feat(codegen): x-expanded-relations

### PR DESCRIPTION
**Pull request by patrick-medusajs**
## What

Alter generated types base on `x-expanded-relations` OAS extension declared on schemaObjects.

## Why

Often, API endpoints will automatically expand a model relations by default. They can also decorate a model with calculated totals. In order to more accurately represent the API, we wish to alter the generated types based on the expanded relations information.

## How

- Follow the relation declaration signature as the backend controllers and the `expand` query param, i.e.: `items.variant.product`.
- Introduce a custom `x-expended-relations` OAS extension.
- Allow for organizing declared relations to help their maintenance.
- Use traversal algorithms in codegen to support deeply nested relationships.
- Use [type-fest](https://www.npmjs.com/package/type-fest)'s `Merge` and `SetRequired` to efficiently alter the types while enabling great intellisense for IDEs.

Extra scope: 
* Added convenience yarn script to interact with the `medisa-oas` CLI within the monorepo. 

## Test

Include in the PR are two implementations of the x-expanded-relations on OAS schema, a simple and a complex one.

### Step 1
* Run `yarn install`
* Run `yarn build`
* Run `yarn medusa-oas oas --type combined --out-dir ~/tmp/oas`
* Run `yarn medusa-oas client --type combined --component types --src-file ~/tmp/oas/combined.osa.json --out-dir ~/tmp/types`
* Open `~/tmp/types/models/StoreRegionsRes`
* Expect relations to be declared as required

### Step 2
* Open `~/tmp/types/models/StoreCartsRes`
* Expect relations to be declared as required
* Expect nested relations to have relations as required.

### Step 3 (optional)
* Open `~/tmp/types` in an intellisense capable IDE
* Within the `index.ts` file, attempt to declare a `const storeRegionRes: StoreRegionRes = {}`
* Expect IDE to highlight that `countries` is a required field of `StoreRegionRes`

### CODE REVIEW

File: packages/medusa/src/api/routes/store/carts/index.ts [hunk 1]
```diff
@@ -154,6 +154,52 @@ export const defaultStoreCartRelations = [
...
+ *     - shipping_methods.shipping_option
+ *   implicit:
+ *     - items.tax_lines
+ *     - items.variant.product
+ *     - region.fulfillment_providers
+ *     - region.payment_providers
+ *     - region.tax_rates
+ *     - shipping_methods.shipping_option
+ *     - shipping_methods.tax_lines
+ *   totals:
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/carts/index.ts:183]: question: Can we make this into a schema? This will be repeated many times I imagine.

File: packages/medusa/src/api/routes/store/carts/index.ts [hunk 2]
```diff
@@ -154,6 +154,52 @@ export const defaultStoreCartRelations = [
...
+ *     - items.variant
+ *     - payment
+ *     - payment_sessions
+ *     - region
+ *     - region.countries
+ *     - region.payment_providers
+ *     - shipping_address
+ *     - shipping_methods
+ *     - shipping_methods.shipping_option
+ *   implicit:
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/carts/index.ts:175]: question: How do you infer implicit relations?

**Review by olivermrbl — COMMENTED** [review]
Great job @patrick-medusajs - it works as expected!

What do you think are good next steps? 

It seems like adding this to all responses is a lot of work, but it might be best to get through it now?

File: packages/medusa/src/api/routes/store/carts/index.ts (hunk 2 above)
Comment (patrick-medusajs) [reply packages/medusa/src/api/routes/store/carts/index.ts:175]: I used `implicit:` to explicitly declare relations that are loaded without the controller's knowledge.

In the case of `StoreCartRes`, we have "eager" loaded relations that are declared on TypeORM models, which I believe we want to move away from in favour of loading them using our expanded relation mechanism. For example, it is the case for `region.payment_providers`. 
https://github.com/medusajs/medusa/blob/831604a5b93797eedd00e8c85a3bd192e597edc2/packages/medusa/src/models/region.ts#L63-L66

The other use case here is to explicitly declare relations that are loaded as part of a subprocess but are not "cleaned" or removed before sending the response. For example, it is the case for `region.tax_rates` that is loaded during the `decorateTotals` process but will not be removed by `defaultStoreCartFields`.
https://github.com/medusajs/medusa/blob/831604a5b93797eedd00e8c85a3bd192e597edc2/packages/medusa/src/services/cart.ts#L2786

In summary, I expect `implicit` to be a temporary measure until implicit relations become explicitly declared in "default" expanded relations or are removed from the response with "default" fields.

File: packages/medusa/src/api/routes/store/carts/index.ts (hunk 1 above)
Comment (patrick-medusajs) [reply packages/medusa/src/api/routes/store/carts/index.ts:183]: There would be a way to optimize but it would be through the introduction of a new entity. Something like `type OrderWithTotals exends Order` and we would use that new entity throughout our codebase as well. 

But I'm not sure we need such an optimization yet. While we have many endpoints that returns models with calculated totals, most of their response shapes are consolidated into only a few schema responses. For example, `StoreCartsRes` in used in 14 endpoints.

**Comment by patrick-medusajs** [comment]
> What do you think are good next steps?

I would merge this feature as is. I would even argue that we could remove the two examples included with it.

Then, I would refactor the OAS for Store responses in one PR and do another PR for Admin responses.

> It seems like adding this to all responses is a lot of work, but it might be best to get through it now?

It's not worst than editing content on a website. It might feel like tedious work but "this is the way". And yes, I would recommend tackling it now.

File: packages/medusa/src/api/routes/store/carts/index.ts (hunk 2 above)
Comment (olivermrbl) [reply packages/medusa/src/api/routes/store/carts/index.ts:175]: Got it - thanks for explaining. Makes a lot of sense. And yes, we are moving away from eager-loaded relations in the near future 👍

File: packages/medusa/src/api/routes/store/carts/index.ts (hunk 1 above)
Comment (olivermrbl) [reply packages/medusa/src/api/routes/store/carts/index.ts:183]: Yes, I agree. Let's keep it as is and consider the totals type at a later point in time :)

**Comment by olivermrbl** [comment]
@patrick-medusajs Sounds like a solid plan 💪
